### PR TITLE
[ios] Telemetry logging

### LINKED ISSUE #3155: Telemetry logging
The iOS SDK should have a debug mode that writes the following information to a file accessible via iTunes:

* Each launch due to a background location update
* Each request the SDK makes against the events API

This will help us better diagnose and verify fixes for bugs such as #1792.

/cc @incanus @friedbunny @bleege @lxbarth

### PR DESCRIPTION
**Pull request by friedbunny**
Adds telemetry logging to a local file (in the app bundle) and to the console.

# SHOULD ONLY BE USED IN INTERNAL DEBUG OR TEST APPS, NEVER IN PRODUCTION

With that disclaimer out of the way:

- Activate by setting `NSUserDefaults` `MGLMapboxMetricsDebugLoggingEnabled`
- To catch all of the events, set this in the app delegate

Fixes #3155

/cc @1ec5 @bleege

### CODE REVIEW

**Comment by tobrun** [comment]
@friedbunny you need to rebase to pass the android test on Bitrise

File: LICENSE.md [hunk 1]
```diff
@@ -885,3 +885,7 @@ freely, subject to the following restrictions:
 
   Jean-loup Gailly        Mark Adler
   jloup@gzip.org          madler@alumni.caltech.edu
+
+===========================================================================
+
+Mapbox GL uses portions of BugshotKit by Marco Arment (copyright 2014), licensed under the MIT License.
```
Comment (1ec5) [inline LICENSE.md]: The [entire text](https://github.com/marcoarment/BugshotKit/blob/master/LICENSE) of the MIT license should be included in this file.

File: ios/app/MBXViewController.mm [hunk 2]
```diff
@@ -144,8 +144,8 @@ - (void)showSettings
                                                        delegate:self
                                               cancelButtonTitle:@"Cancel"
                                          destructiveButtonTitle:nil
-                                              otherButtonTitles:@"Reset North",
-                                                                @"Reset Position",
+                                              otherButtonTitles:@"Print telemetry logfile",
+                                                                @"Delete telemetry logfile",
```
Comment (1ec5) [inline ios/app/MBXViewController.mm]: Title Case

File: ios/app/MBXViewController.mm [hunk 3]
```diff
@@ -144,8 +144,8 @@ - (void)showSettings
                                                        delegate:self
                                               cancelButtonTitle:@"Cancel"
                                          destructiveButtonTitle:nil
-                                              otherButtonTitles:@"Reset North",
-                                                                @"Reset Position",
```
Comment (1ec5) [inline ios/app/MBXViewController.mm]: Reset Position can be useful, but I agree with nixing Reset North.

File: LICENSE.md (hunk 1 above)
Comment (friedbunny) [inline LICENSE.md]: Mrrrrrgh, legal is as legal does and no one reads these anyway, but the MIT license is included twice in this document already.

File: platform/ios/src/MGLMapboxEvents.m [hunk 4]
```diff
@@ -880,7 +930,133 @@ - (void)URLSession:(NSURLSession *)session didReceiveChallenge:(NSURLAuthenticat
...
+    NSData *plistdata_latin1 = [plistString dataUsingEncoding:NSISOLatin1StringEncoding];
+    NSError *error = nil;
+    NSDictionary *mobileProvision = [NSPropertyListSerialization propertyListWithData:plistdata_latin1 options:NSPropertyListImmutable format:NULL error:&error];
+    if (error) return YES; // unknown plist format
+
+    if (! mobileProvision || ! mobileProvision.count) return YES; // no entitlements
+
+    if (mobileProvision[@"ProvisionsAllDevices"]) return NO; // enterprise provisioning
+
+    if (mobileProvision[@"ProvisionedDevices"] && ((NSDictionary *)mobileProvision[@"ProvisionedDevices"]).count) return NO; // development or ad-hoc
```
Comment (1ec5) [inline platform/ios/src/MGLMapboxEvents.m]: Nit: You can avoid this ugly cast by using the bracketed message syntax: `[mobileProvision[@"ProvisionedDevices"] count]`.

**Comment by 1ec5** [comment]
Does this need to be on the 3.1.0 milestone? It shouldn’t affect shipping code anyways.

**Comment by friedbunny** [comment]
Rebased and cleaned up! @1ec5 I'd like to get this in v3.1.0, but it's not a huuuUUUuuuge priority, certainly.

**Comment by tobrun** [comment]
@1ec5, there was a problem with a proguard configuration at that time and thus breaking the build. (unrelated to release of 3.1.0 or the branch)

**Comment by friedbunny** [comment]
Squashed and rebased.

**Comment by 1ec5** [comment]
:+1:

**Comment by friedbunny** [comment]
Merged this as 4a3adddf3f4f833768f680d2d1e228c88f2e06e3 and f9c9ee93515ab26a508d0dec7a48cdab5fb70500. Pushed to `master` before force-pushing this branch, which confused Github. :ghost:

**Pull request closed without merging**